libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SdkVersion.java

```diff
@@ -0,0 +1,149 @@
+package com.salesforce.androidsdk.app;
+
+import androidx.annotation.NonNull;
+
+import java.util.regex.Pattern;
+
+public final class SdkVersion implements Comparable<SdkVersion> {
+    private final int major;
+    private final int minor;
+    private final int patch;
+    private final boolean isDev;
+
+    public SdkVersion(int major, int minor, int patch, boolean isDev) throws IllegalArgumentException {
+        if (major < 0 || minor < 0 || patch < 0) {
+            throw new IllegalArgumentException("Invalid version number combination: major=" + major + ", minor=" + minor + ", patch=" + patch);
+        }
+
+        this.major = major;
+        this.minor = minor;
+        this.patch = patch;
+        this.isDev = isDev;
+    }
+
+    public int getMajor() {
+        return major;
+    }
+
+    public int getMinor() {
+        return minor;
+    }
+
+    public int getPatch() {
+        return patch;
+    }
+
+    public boolean isDev() {
+        return isDev;
+    }
+
+    public boolean isGreaterThan(@NonNull final SdkVersion o) {
+        return this.compareTo(o) > 0;
+    }
+
+    public boolean isGreaterThanOrEqualTo(@NonNull final SdkVersion o) {
+        return this.compareTo(o) >= 0;
+    }
+
+    public boolean isLessThan(@NonNull final SdkVersion o) {
+        return this.compareTo(o) < 0;
+    }
+
+    public boolean isLessThanOrEqualTo(@NonNull final SdkVersion o) {
+        return this.compareTo(o) <= 0;
+    }
+
+    /**
+     * Matches version strings in the form of XX.YY.ZZ[.dev], where each version number can be 1-9 digits long.
+     */
+    private static final Pattern VERSION_STR_PATTERN = Pattern.compile("^\\d{1,9}\\.\\d{1,9}\\.\\d{1,9}(\\.dev)?$");
+
+    public static SdkVersion parseFromString(@NonNull final String versionStr) throws IllegalArgumentException {
+        final String trimmed = versionStr.trim();
+        if (!VERSION_STR_PATTERN.matcher(trimmed).matches()) {
+            throw new IllegalArgumentException("Version string \"" + trimmed + "\" did not match expected pattern of XX.YY.ZZ[.dev]");
+        }
+
+        final String[] parts = trimmed.split("\\.");
+        return new SdkVersion(
+                Integer.parseInt(parts[0]),
+                Integer.parseInt(parts[1]),
+                Integer.parseInt(parts[2]),
+                parts.length == 4 // If we pass regex matching then we know the final part is "dev"
+        );
+    }
+
+    @Override
+    public int compareTo(SdkVersion o) {
```

Comment:
This method feels like it could be compacted, but this is actually the most optimal algorithm for comparing two versions.  First check the major version, then minor, etc.